src/analysis/normed/group/seminorm.lean

```diff
@@ -0,0 +1,328 @@
+
+attribute [nolint doc_blame] add_group_seminorm.to_zero_hom
+
+/-- `add_group_seminorm_class F α` states that `F` is a type of seminorms on the additive group `α`.
+
+You should extend this class when you extend `add_group_seminorm`. -/
+class add_group_seminorm_class (F : Type*) (α : out_param $ Type*) [add_group α]
+  extends fun_like F α (λ _, ℝ) :=
+(map_zero (f : F) : f 0 = 0)
+(map_add_le_add (f : F) (a b : α) : f (a + b) ≤ f a + f b)
+(map_neg_eq_map (f : F) (a : α) : f (-a) = f a)
+
+/-- `group_seminorm_class F α` states that `F` is a type of seminorms on the group `α`.
+
+You should extend this class when you extend `group_seminorm`. -/
+@[to_additive]
+class group_seminorm_class (F : Type*) (α : out_param $ Type*) [group α]
+  extends fun_like F α (λ _,  ℝ) :=
+(map_one_eq_zero (f : F) : f 1 = 0)
+(map_mul_le_add (f : F) (a b : α) : f (a * b) ≤ f a + f b)
+(map_inv_eq_map (f : F) (a : α) : f a⁻¹ = f a)
+
+export add_group_seminorm_class (map_add_le_add map_neg_eq_map)
+export group_seminorm_class (map_one_eq_zero map_mul_le_add map_inv_eq_map)
+
+attribute [simp, to_additive map_zero] map_one_eq_zero
+attribute [simp] map_neg_eq_map
+attribute [simp, to_additive] map_inv_eq_map
+
+@[priority 100] -- See note [lower instance priority]
+instance add_group_seminorm_class.to_zero_hom_class [add_group E] [add_group_seminorm_class F E] :
+  zero_hom_class F E ℝ :=
+{ ..‹add_group_seminorm_class F E› }
+
+section group
+variables [group E] [group_seminorm_class F E] (f : F) (x y : E)
+include E
+
+@[to_additive] lemma map_div_le_add : f (x / y) ≤ f x + f y :=
+by { rw [div_eq_mul_inv, ←map_inv_eq_map f y], exact map_mul_le_add _ _ _ }
+
+@[simp, to_additive] lemma map_nonneg_mul : 0 ≤ f x :=
+nonneg_of_mul_nonneg_right
+  (by { rw [two_mul, ←map_one_eq_zero f, ←div_self' x], exact map_div_le_add _ _ _ }) two_pos
+
+@[to_additive] lemma map_div_rev : f (x / y) = f (y / x) := by rw [←inv_div, map_inv_eq_map]
+
+/-- The direct path from `1` to `x` is shorter than the path with `y` "inserted" in between. -/
+@[to_additive "The direct path from `0` to `x` is shorter than the path with `y` \"inserted\" in
+between."]
+lemma le_map_add_map_div' : f x ≤ f y + f (x / y) :=
+by simpa only [add_comm, div_mul_cancel'] using map_mul_le_add f (x / y) y
+
+/-- The direct path from `1` to `y` is shorter than the path with `x` "inserted" in between. -/
+@[to_additive "The direct path from `0` to `y` is shorter than the path with `x` \"inserted\" in
+between."]
+lemma le_map_add_map_div : f y ≤ f x + f (x / y) :=
+by simpa only [add_comm, map_div_rev, div_mul_cancel'] using map_mul_le_add f (y / x) x
+
+end group
+
+namespace tactic
+open positivity
+
+/-- Extension for the `positivity` tactic: seminorms are nonnegative. -/
+@[positivity]
+meta def positivity_seminorm : expr → tactic strictness
+| (expr.app `(⇑%%f) `(%%a)) := nonnegative <$> mk_app ``map_nonneg_add [f, a]
+               <|> nonnegative <$> mk_app ``map_nonneg_mul [f, a]
```

Comment:
This seems like it might be too universally applicable a match, similar to the [one with canonically ordered monoids](https://github.com/leanprover-community/mathlib/pull/16141#discussion_r953731181) which @Vierkantor and @digama0 raised concerns about in #16141.  I don't feel confident judging; I'll let them comment.

Reply:
I think having the head symbol being an application + the second head symbol being `coe_fn` significantly reduces the likeliness of a match.

Reply:
I think this is probably fine: `map_nonneg` is already a `simp` lemma so this should have less impact than that.